src/v/storage/segment_set.cc

```diff
@@ -437,6 +437,14 @@ static ss::future<segment_set> unsafe_do_recover(
             vlog(stlog.info, "Recovered: {}", s);
             good.emplace_back(std::move(s));
         }
+
+        // Pass `good` by value on purpose, we need to preserve segments for
+        // sad return path below in case we are unable to produce a
+        // contiguous segment set here.
+        auto seg_set_opt = maybe_create_contiguous_segment_set(good).get();
+        if (seg_set_opt.has_value()) {
+            return std::move(seg_set_opt).value();
+        }
         return segment_set(std::move(good));
```

Comment:
> if this function fails to create a contiguous non-overlapping set of `segment`s to represent the `log` with, the original `deque` of `segment`s is returned instead (i.e the old behavior is preserved).

So we are accepting running with a segment set which does not make sense and results pretty much in undefined behavior? Why this instead of failing to start the partition and asking the operator to intervene?

Reply:
Yeah, my rationale was that the edge case Willem has been concerned about at is actually pretty narrow -- unclean shutdown during compaction. That seemed easy enough to provide a targeted fix with somewhat of a guarantee that we don't regress anything else.

I'm not against doing the broader change too, but I don't think it needs to block this change, and may warrant a bit more testing and bake time